### PR TITLE
Fix: EarlyStopping does not store and restore the model #175

### PR DESCRIPTION
**Pull request by LuoXueling**
See issue #175

### CODE REVIEW

**Comment by LuoXueling** [comment]
Explanation of the unit test:

`min_delta=1000`, so `EarlyStopping` always stores the `state_dict()` of the first epoch. `patience=1000` is not necessary here because `n_epochs=2` but I just want to mention that, if only the deep copy thing is fixed, the bug also exists if early stopping is not triggered. Finally, the restored model should have the same validation loss as that in the first epoch.

**Comment by kd1510** [comment]
Thanks for putting this together - will have time to review tomorrow if nobody else gets around to it!
